integration/spark/build.gradle

```diff
@@ -104,6 +102,20 @@ dependencies {
     testCompile "org.mockito:mockito-core:${mockitoVersion}"
     testCompile "org.mockito:mockito-junit-jupiter:${mockitoVersion}"
     testCompileOnly "org.projectlombok:lombok:${lombokVersion}"
+    testCompileOnly "org.apache.spark:spark-core_2.11:${sparkVersion}"
+    testCompileOnly "org.apache.spark:spark-sql_2.11:${sparkVersion}"
+    testCompileOnly 'com.google.cloud.spark:spark-bigquery-with-dependencies_2.11:0.21.1'
+    if (project.getProperty('spark.version').startsWith('3')) {
+        runtimeOnly "com.fasterxml.jackson.module:jackson-module-scala_2.12:${jacksonVersion}"
+        testRuntimeOnly 'com.google.cloud.spark:spark-bigquery-with-dependencies_2.12:0.21.1'
+        testRuntimeOnly "org.apache.spark:spark-core_2.12:${spark3Version}"
+        testRuntimeOnly "org.apache.spark:spark-sql_2.12:${spark3Version}"
+    } else {
+        runtimeOnly "com.fasterxml.jackson.module:jackson-module-scala_2.11:${jacksonVersion}"
+        testRuntimeOnly 'com.google.cloud.spark:spark-bigquery-with-dependencies_2.11:0.21.1'
+        testRuntimeOnly "org.apache.spark:spark-core_2.11:${sparkVersion}"
+        testRuntimeOnly "org.apache.spark:spark-sql_2.11:${sparkVersion}"
+    }
```

Comment:
We should be able to do this with Gradle build `configurations`.  See https://docs.gradle.org/6.8.1/userguide/java_testing.html#sec:configuring_java_integration_tests

Define a `configurations` block like
```
configurations {
    spark3.extendsFrom compileOnly
    spark2.extendsFrom compileOnly
    spark3Test.extendsFrom testImplementation
    spark2Test.extendsFrom testImplementation
}
```
Configure each of the dependencies with the appropriate configuration label. Then, in the `test` block, you can update the classpath with the appropriate configuration's classpath:

```
test {
    useJUnitPlatform {
        excludeTags 'integration-test'
    }
    classpath = configurations.spark3.runtimeClasspath
    // ...
```

This gets us closer to running compilation in multiple steps with different sets of dependencies.

Reply:
so, you want to have separate test tasks
for spark 2.4.7
```
test {
...
}
integrationTest {
}

and for spark 3
```
testSpark3 {
...
}
integrationTestSpark3 {
}
```
Cause I wanted to have only two of them
```
test {
...
}
integrationTest {
}
```
and in CI add param to build it for spark3

Reply:
@collado-mike what do you think if we leave only two tasks and in CI we will use matrix-job, as @wslulciuc  suggests.
If we want to have separate tasks, we can use it as:

```
- run: ./gradlew --no-daemon --stacktrace << parameters.gradle-task >>

...
parameters:
              gradle-task: [test, testSpark3]
```
vs
```
- run: ./gradlew --no-daemon --stacktrace test -Pspark.version=<< parameters.spark-version >>
...
parameters:
     spark-version: [2.4.7, 3.1]
```